common/logger/src/lib.rs

```diff
@@ -1,74 +1,426 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-#![forbid(unsafe_code)]
-
-use chrono::Utc;
+pub use chrono::Utc;
 use env_logger::filter;
 pub use log::{self, Level, Log, Metadata, Record};
+pub use serde_json;
+
+use serde::Serialize;
+use serde_json::Value;
 use std::{
+    collections::HashMap,
     env,
     fmt::{self, Write},
-    sync::mpsc::{self, Receiver, RecvError, SyncSender, TrySendError},
+    fs::{File, OpenOptions},
+    io,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        mpsc::{self, Receiver, RecvError, SyncSender, TrySendError},
+    },
     thread,
 };
 
 pub mod prelude {
     pub use crate::{crit, debug, error, info, trace, warn};
 }
+pub use std::io::Write as IoWrite;
 
 pub const CHANNEL_SIZE: usize = 256;
 pub const DEFAULT_TARGET: &str = "libra";
 const RUST_LOG: &str = "RUST_LOG";
 
+pub trait StructLogSink: Sync {
```

Comment:
maybe we should separate out this file into two modules:
structured.rs
unstructured.rs
and maybe
macros.rs

and only include here our re-exports?

Reply:
Even with comments/documentation this lib.rs is still under 1000 lines.
For comparison, rust log crate has lib.rs of 1.6k lines. Is it really necessary to split?

Reply:
My personal preference is to split based upon the level of modularity and only merge multiple modules if it ends up being < 100 ~ 200 lines of code. This will effectively contain 3 different modules and 1 major feature (structured logging), so it seems relatively straightforward on how one would break it up into pieces. We are also likely going to extend upon the complexity within this crate.

With respect to the rust log crate, I did not enjoy my time with that crate. I spent far more time searching side of it (via ctrl-f) than I would like to admit. Ultimately it made me better at navigating large, single files within Rust, but definitely not a flavor I prefer.

I'll defer to you, but I would really like to see this split up.